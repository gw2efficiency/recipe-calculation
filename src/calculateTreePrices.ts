import { RecipeTreeWithCraftFlags, RecipeTreeWithPrices, RecipeTreeWithQuantity } from './types'
import { CURRENCY_DECISION_PRICES } from './static/currencyDecisionPrices'

// Update the tree prices
export function calculateTreePrices(
  tree: RecipeTreeWithCraftFlags,
  itemPrices: Record<string, number>,
  customCurrencyPrices: Record<string, number>
): RecipeTreeWithCraftFlags
export function calculateTreePrices(
  tree: RecipeTreeWithQuantity,
  itemPrices: Record<string, number>,
  customCurrencyPrices: Record<string, number>
): RecipeTreeWithPrices
export function calculateTreePrices(
  tree: RecipeTreeWithQuantity | RecipeTreeWithCraftFlags,
  itemPrices: Record<string, number>,
  customCurrencyPrices: Record<string, number>
): RecipeTreeWithPrices | RecipeTreeWithCraftFlags {
  // Calculate the buy prices
  let buyPriceEach = itemPrices[tree.id] || false
  if (tree.type === 'Currency') {
    // Coin is 1c per used quantity, the other currencies do not have a TP price
    buyPriceEach = tree.id === 1 ? 1 : false
  }
  const buyPrice = buyPriceEach ? tree.usedQuantity * buyPriceEach : false
  let craftResultPrice = buyPrice

  // Calculate the base decision price (may be overwritten later with the craft price)
  let decisionPriceEach = buyPriceEach || undefined
  if (tree.type === 'Currency') {
    if (customCurrencyPrices && typeof customCurrencyPrices[tree.id] === 'number') {
      decisionPriceEach = customCurrencyPrices[tree.id]
    } else {
      decisionPriceEach = CURRENCY_DECISION_PRICES[tree.id]
    }
  }
  let decisionPrice = decisionPriceEach ? tree.usedQuantity * decisionPriceEach : false

  if (!tree.components) {
    return {
      ...tree,
      components: undefined,
      buyPriceEach,
      buyPrice,
      decisionPrice,
      craftResultPrice,
      craftDecisionPrice: decisionPrice,
    }
  }

  // Calculate the tree prices traversal for the sub-components
  const components = tree.components.map((component) =>
    calculateTreePrices(component, itemPrices, customCurrencyPrices)
  )

  // Calculate the craft price out of the best prices
  const craftDecisionPrice = components.map((c) => c.decisionPrice || 0).reduce((a, b) => a + b, 0)
  const craftPrice = components.map((c) => c.craftResultPrice || 0).reduce((a, b) => a + b, 0)

  // Update the decision price of this tree segment to the craft price,
  // used to determine the craft price of the higher-up recipe
  if (
    !('craft' in tree && tree.craft === false) &&
    (('craft' in tree && tree.craft === true) ||
      !decisionPrice ||
      craftDecisionPrice < decisionPrice)
  ) {
    decisionPrice = craftDecisionPrice
    craftResultPrice = craftPrice
  }

  craftResultPrice = craftResultPrice || craftPrice
  decisionPrice = decisionPrice || craftDecisionPrice

  return {
    ...tree,
    components,
    buyPriceEach,
    buyPrice,
    craftPrice,
    decisionPrice,
    craftResultPrice,
    craftDecisionPrice,
  }
}
