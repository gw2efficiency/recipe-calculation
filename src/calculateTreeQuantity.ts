import { RecipeTreeWithCraftFlags, RecipeTree, RecipeTreeWithQuantity } from './types'

export function calculateTreeQuantity(
  amount: number,
  tree: RecipeTreeWithCraftFlags,
  availableItems?: Record<string, number>
): RecipeTreeWithCraftFlags
export function calculateTreeQuantity(
  amount: number,
  tree: RecipeTree,
  availableItems?: Record<string, number>
): RecipeTreeWithQuantity
export function calculateTreeQuantity(
  amount: number,
  tree: RecipeTree | RecipeTreeWithCraftFlags,
  availableItems: Record<string, number> = {}
) {
  // Make sure that we don't modify the passed-in object
  // We still want to work with a reference in the actual calculation
  // since the availableItems are a shared state for all sub-recipes
  return calculateTreeQuantityInner(amount, tree, { ...availableItems })
}

// Go through a recipe tree and set 'totalQuantity' based on the
// wanted amount and the output of recipes and sub-recipes
function calculateTreeQuantityInner(
  amount: number,
  tree: RecipeTree | RecipeTreeWithCraftFlags,
  availableItems: Record<string, number>,
  ignoreAvailable = false,
  nesting = 0
): RecipeTreeWithCraftFlags | RecipeTreeWithQuantity {
  const output = tree.output || 1

  // Calculate the total quantity needed
  let treeQuantity = amount * tree.quantity

  // Round amount to nearest multiple of the tree output
  treeQuantity = Math.ceil(treeQuantity / output) * output
  const totalQuantity = Math.round(treeQuantity)

  // If the item is available and the higher tree is not
  // bought or already available get as many items of it as possible
  // (This ignores the root node, because we *always* want to craft all of these)
  let availableQuantity = 0
  if (nesting > 0 && tree.type !== 'Currency' && !ignoreAvailable && availableItems[tree.id]) {
    availableQuantity = Math.min(availableItems[tree.id], totalQuantity)
    availableItems[tree.id] -= availableQuantity
  }
  const usedQuantity = totalQuantity - availableQuantity

  if (!tree.components) {
    return { ...tree, components: undefined, output, totalQuantity, usedQuantity }
  }

  // Get the amount of components that need to be crafted
  // e.g. a recipe outputs 10 and we need 20 -> 2x components
  const componentAmount = Math.ceil(usedQuantity / output)

  // Ignore available items in components if the tree
  // doesn't get crafted or is completely available anyway
  ignoreAvailable =
    ('craft' in tree && tree.craft === false) || usedQuantity === 0 || ignoreAvailable

  // Adjust the quantity for all tree's subcomponents
  const components = tree.components.map((component) => {
    return calculateTreeQuantityInner(
      componentAmount,
      component,
      availableItems,
      ignoreAvailable,
      ++nesting
    )
  })

  return { ...tree, components, output, totalQuantity, usedQuantity }
}
