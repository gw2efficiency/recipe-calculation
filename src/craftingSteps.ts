import { Prerequisites } from '@gw2efficiency/recipe-nesting'
import { RecipeTreeWithCraftFlags } from './types'

export function craftingSteps(tree: RecipeTreeWithCraftFlags) {
  let steps = craftingStepsInner(tree).reverse()

  // We don't care about steps where we have to craft "0" of something
  // This can happen when a crafting tree is updated after the initial calculation
  steps = steps.filter((step) => step.quantity > 0)

  // Make sure that "Mystic Clovers" (and it's components) get crafted as the first steps,
  // since they generate items that are always useful for crafting the other steps
  // Obsidian Shards and Philosopher Stones (required for Mystic Clovers) will be sorted
  // above the Mystic Clovers in the next step when sorting merchants
  const mysticCloversId = 19675
  const mysticClovers = steps.find((step) => step.id === mysticCloversId)
  if (mysticClovers) {
    steps = steps.filter((step) => step.id !== mysticCloversId)
    steps.unshift(mysticClovers)
  }

  // Sort merchants that only require currencies to the top
  steps = steps.sort((a, b) => {
    const aIsMerchant = isMerchantWithOnlyCurrencies(a)
    const bIsMerchant = isMerchantWithOnlyCurrencies(b)

    return aIsMerchant === bIsMerchant ? 0 : aIsMerchant ? -1 : 1
  })

  return steps.map((step) => ({
    ...step,
    // Calculate how many times you actually have to click on "craft"
    // for items with output > 1 (calculate here when all steps are aggregated)
    crafts: Math.ceil(step.quantity / step.output),
    output: undefined,
  }))
}

interface CraftingStep {
  id: number
  type: 'Item' | 'Recipe' | 'Currency'
  output: number
  quantity: number
  minRating: number | null
  disciplines: Array<string>
  merchant?: { name: string; locations: Array<string> }
  prerequisites: Prerequisites
  components: Array<{ id: number; type: 'Item' | 'Recipe' | 'Currency'; quantity: number }>
}

// Generate an ordered list of crafting steps
function craftingStepsInner(
  tree: RecipeTreeWithCraftFlags,
  steps: Array<CraftingStep> = [],
  index = 0
) {
  const treeComponents = tree.components

  // Skip any tree parts where nothing needs to be crafted
  if (!treeComponents || tree.craft === false || tree.type === 'Currency') {
    return steps
  }

  // Go through the existing steps, and if we already have a step
  // with this id, just add up the quantities
  const stepIndex = steps.findIndex((step) => step.id === tree.id)
  if (stepIndex !== -1) {
    steps[stepIndex].quantity += tree.usedQuantity
    steps[stepIndex].components = steps[stepIndex].components.map((component) => {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const treeComponent = treeComponents.find((x) => x.id === component.id)!

      component.quantity += treeComponent.totalQuantity
      return component
    })
    index = stepIndex
  }

  // We don't have a step like this yet, push a new one at the given index
  if (stepIndex === -1) {
    steps.splice(index, 0, {
      id: tree.id,
      type: tree.type,
      output: tree.output,
      quantity: tree.usedQuantity,
      minRating: tree.min_rating,
      disciplines: tree.disciplines,
      merchant: tree.merchant,
      prerequisites: tree.prerequisites,
      components: treeComponents.map((component) => ({
        id: component.id,
        type: component.type,
        quantity: component.totalQuantity,
      })),
    })
  }

  // Go through the components and push them after the index of their parent
  treeComponents.map((component) => craftingStepsInner(component, steps, index + 1))
  return steps
}

function isMerchantWithOnlyCurrencies(step: CraftingStep): boolean {
  return (
    step.disciplines.length === 1 &&
    step.disciplines[0] === 'Merchant' &&
    step.components.every(({ type }) => type === 'Currency')
  )
}
