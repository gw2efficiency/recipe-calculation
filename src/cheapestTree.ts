import { calculateTreeQuantity } from './calculateTreeQuantity'
import { calculateTreePrices } from './calculateTreePrices'
import { calculateTreeCraftFlags } from './calculateTreeCraftFlags'
import { RecipeTree, RecipeTreeWithCraftFlags, RecipeTreeWithPrices } from './types'
import {
  NestedRecipe,
  BasicCurrencyComponent,
  BasicItemComponent,
} from '@gw2efficiency/recipe-nesting'

export function cheapestTree(
  amount: number,
  tree: NestedRecipe,
  itemPrices: Record<string, number>,
  availableItems: Record<string, number> = {},
  forceBuyItems: Array<number> = [],
  valueOwnItems = false,
  userEfficiencyTiers: Record<string, string> = {
    '102306': '0',
    '102205': '0',
    '103049': '0',
  },
  customCurrencyPrices: Record<string, number>
): RecipeTreeWithCraftFlags {
  tree = applyEfficiencyTiersToTree(tree, userEfficiencyTiers)
  if (valueOwnItems) {
    const treeWithQuantityWithoutAvailableItems = calculateTreeQuantity(
      amount,
      tree as RecipeTree,
      {}
    )
    const treeWithPriceWithoutAvailableItems = calculateTreePrices(
      treeWithQuantityWithoutAvailableItems,
      itemPrices,
      customCurrencyPrices
    )

    const cheaperToBuyItemIds = getCheaperToBuyItemIds(treeWithPriceWithoutAvailableItems)

    // calculateTreeQuantity already checks for craft flags, so we can set them here
    tree = disableCraftForItemIds(tree, cheaperToBuyItemIds) as NestedRecipe
  }

  // Adjust the tree total and used quantities
  const treeWithQuantity = calculateTreeQuantity(amount, tree as RecipeTree, availableItems)

  // Set the initial craft flags based on the subtree prices
  const treeWithPrices = calculateTreePrices(treeWithQuantity, itemPrices, customCurrencyPrices)
  let treeWithCraftFlags = calculateTreeCraftFlags(treeWithPrices, forceBuyItems)

  // Force the root to be crafted
  treeWithCraftFlags = { ...treeWithCraftFlags, craft: true }

  // After the "craft" flags are set, update the used materials
  // to only be used for things that actually get crafted
  const treeWithQuantityPostFlags = calculateTreeQuantity(
    amount,
    treeWithCraftFlags,
    availableItems
  )

  // Recalculate the correct tree price
  return calculateTreePrices(treeWithQuantityPostFlags, itemPrices, customCurrencyPrices)
}

function getCheaperToBuyItemIds(
  tree: RecipeTreeWithPrices,
  ids: Array<number> = []
): Array<number> {
  // If buying the item costs less than the possible profit made by selling its craft components (with TP tax), we should buy instead of crafting
  if (
    typeof tree.craftDecisionPrice === 'number' &&
    typeof tree.buyPrice === 'number' &&
    tree.buyPrice < tree.craftDecisionPrice * 0.85
  ) {
    if (!ids.includes(tree.id)) {
      ids.push(tree.id)
    }
  }

  if (tree.components && Array.isArray(tree.components)) {
    tree.components.forEach((component) => getCheaperToBuyItemIds(component, ids))
  }

  return ids
}

type NestedRecipeAndBasicComponentWithCraftFlag = (
  | NestedRecipe
  | BasicItemComponent
  | BasicCurrencyComponent
) & {
  craft?: boolean
}

function disableCraftForItemIds(
  tree: NestedRecipeAndBasicComponentWithCraftFlag,
  ids: Array<number>
) {
  if (ids.includes(tree.id)) {
    tree = { ...tree, craft: false }
  }

  if ('components' in tree && Array.isArray(tree.components)) {
    tree.components = tree.components.map((component) =>
      disableCraftForItemIds(component as NestedRecipeAndBasicComponentWithCraftFlag, ids)
    )
  }

  return tree
}

function applyEfficiencyTiersToTree(
  tree: Omit<NestedRecipe, 'id'> & { id: number | null }, // FIXME Not sure why this can be null
  userEfficiencyTiers: Record<string, string>
): NestedRecipe {
  const id = tree.id ? tree.id.toString() : ''

  if (
    ['102306', '102205', '103049'].includes(id) &&
    tree.merchant &&
    tree.merchant.name.includes('Homestead Refinement')
  ) {
    const efficiencyTier = Number(userEfficiencyTiers[id])

    if (efficiencyTier > 0) {
      const component = { ...tree.components[0] }

      // Each efficiency tier lowers input by 50%, if it drops below one then doubles output
      component.quantity = component.quantity / (efficiencyTier * 2)

      // Bug: Onions are discounted by 75% with first tier
      if (component.id === 12142) {
        component.quantity = efficiencyTier === 1 ? 1 : 0.5
      }

      // Bug: Potatoes are not discounted with first tier
      if (component.id === 12135) {
        component.quantity = efficiencyTier === 1 ? 8 : 4
      }

      let updatedTree = { ...tree, output: component.quantity < 1 ? tree.output * 2 : tree.output }

      // Bug: Iron ore output also halves with second tier
      if (component.id === 19699 && efficiencyTier === 2) {
        updatedTree.output = updatedTree.output / 2
      }

      component.quantity = component.quantity < 1 ? 1 : component.quantity
      updatedTree = { ...updatedTree, components: [component, ...tree.components.slice(1)] }

      tree = updatedTree
    }
  }

  if ('components' in tree && Array.isArray(tree.components)) {
    tree = {
      ...tree,
      components: tree.components.map((component) =>
        applyEfficiencyTiersToTree(component as NestedRecipe, userEfficiencyTiers)
      ),
    }
  }

  return tree as NestedRecipe
}
