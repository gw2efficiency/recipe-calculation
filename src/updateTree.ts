import { calculateTreeQuantity } from './calculateTreeQuantity'
import { calculateTreePrices } from './calculateTreePrices'
import { RecipeTreeWithCraftFlags } from './types'

export function updateTree(
  amount: number,
  tree: RecipeTreeWithCraftFlags,
  itemPrices: Record<string, number>,
  availableItems: Record<string, number> = {},
  customCurrencyPrices: Record<string, number> = {}
) {
  // Update the tree total and used quantities
  const treeWithQuantity = calculateTreeQuantity(amount, tree, availableItems)

  // Recalculate the correct tree price
  return calculateTreePrices(treeWithQuantity, itemPrices, customCurrencyPrices)
}
