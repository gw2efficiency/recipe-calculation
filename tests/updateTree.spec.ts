import { clone } from '@devoxa/flocky'
import { RecipeTreeWithCraftFlags } from '../src/types'
import { updateTree } from '../src'

describe('updateTree', () => {
  const calculatedTree: RecipeTreeWithCraftFlags = {
    craft: true,
    craftDecisionPrice: 384,
    craftPrice: 384,
    id: 1,
    output: 1,
    quantity: 1,
    totalQuantity: 2,
    usedQuantity: 2,
    decisionPrice: 384,
    craftResultPrice: 384,
    buyPrice: 20,
    buyPriceEach: 10,
    type: 'Recipe',
    min_rating: 123,
    disciplines: ['Chef'],
    multipleRecipeCount: 1,
    daily_purchase_cap: 0,
    weekly_purchase_cap: 0,
    components: [
      {
        craft: false,
        craftDecisionPrice: 84,
        decisionPrice: 84,
        craftResultPrice: 84,
        buyPrice: 84,
        buyPriceEach: 42,
        id: 2,
        output: 1,
        quantity: 1,
        totalQuantity: 2,
        usedQuantity: 2,
        type: 'Item',
        min_rating: null,
        disciplines: [],
        prerequisites: [],
        multipleRecipeCount: 1,
        daily_purchase_cap: 0,
        weekly_purchase_cap: 0,
      },
      {
        craft: false,
        craftDecisionPrice: 200,
        decisionPrice: 100,
        craftResultPrice: 100,
        buyPrice: 100,
        buyPriceEach: 10,
        craftPrice: 200,
        id: 3,
        output: 1,
        quantity: 5,
        totalQuantity: 10,
        usedQuantity: 10,
        type: 'Recipe',
        min_rating: null,
        disciplines: [],
        multipleRecipeCount: 1,
        daily_purchase_cap: 0,
        weekly_purchase_cap: 0,
        components: [
          {
            craft: false,
            craftDecisionPrice: 200,
            decisionPrice: 200,
            craftResultPrice: 200,
            buyPrice: 200,
            buyPriceEach: 10,
            id: 4,
            output: 1,
            quantity: 2,
            totalQuantity: 20,
            usedQuantity: 20,
            type: 'Item',
            min_rating: null,
            disciplines: [],
            prerequisites: [],
            multipleRecipeCount: 1,
            daily_purchase_cap: 0,
            weekly_purchase_cap: 0,
          },
        ],
        prerequisites: [],
      },
      {
        craft: true,
        craftDecisionPrice: 200,
        decisionPrice: 200,
        craftResultPrice: 200,
        buyPrice: 10000,
        buyPriceEach: 1000,
        craftPrice: 200,
        id: 5,
        output: 1,
        quantity: 5,
        totalQuantity: 10,
        usedQuantity: 10,
        type: 'Recipe',
        min_rating: null,
        disciplines: [],
        multipleRecipeCount: 1,
        daily_purchase_cap: 0,
        weekly_purchase_cap: 0,
        components: [
          {
            craft: false,
            craftDecisionPrice: 200,
            decisionPrice: 200,
            craftResultPrice: 200,
            buyPrice: 200,
            buyPriceEach: 10,
            id: 6,
            output: 1,
            quantity: 2,
            totalQuantity: 20,
            usedQuantity: 20,
            type: 'Item',
            min_rating: null,
            disciplines: [],
            prerequisites: [],
            multipleRecipeCount: 1,
            daily_purchase_cap: 0,
            weekly_purchase_cap: 0,
          },
        ],
        prerequisites: [],
      },
    ],
    prerequisites: [],
  }

  it('keeps the tree the same if nothing changed', () => {
    const tree = clone(calculatedTree)
    const prices = { 1: 10, 2: 42, 3: 10, 4: 10, 5: 1000, 6: 10 }
    const updatedTree = updateTree(2, tree, prices)
    expect(updatedTree).toEqual(calculatedTree)
  })

  it('updates a tree correctly if the amount, prices or craft flags changed', () => {
    const tree = clone(calculatedTree)
    tree.components![1].craft = true // eslint-disable-line @typescript-eslint/no-non-null-assertion
    const prices = { 1: 10, 2: 42, 3: 10, 4: 10, 5: 1000, 6: 11 }
    const updatedTree = updateTree(5, tree, prices)
    expect(updatedTree).toMatchSnapshot()
  })

  it('updates a tree correctly if the available items changed', () => {
    const tree = clone(calculatedTree)
    tree.components![1].craft = true // eslint-disable-line @typescript-eslint/no-non-null-assertion
    const prices = { 1: 10, 2: 42, 3: 10, 4: 10, 5: 1000, 6: 11 }
    const updatedTree = updateTree(5, tree, prices, { 2: 1000 })
    expect(updatedTree).toMatchSnapshot()
  })
})
