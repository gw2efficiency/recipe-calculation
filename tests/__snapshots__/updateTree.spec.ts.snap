// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateTree updates a tree correctly if the amount, prices or craft flags changed 1`] = `
Object {
  "buyPrice": 50,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 210,
      "buyPriceEach": 42,
      "components": undefined,
      "craft": false,
      "craftDecisionPrice": 210,
      "craftResultPrice": 210,
      "daily_purchase_cap": 0,
      "decisionPrice": 210,
      "disciplines": Array [],
      "id": 2,
      "min_rating": null,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [],
      "quantity": 1,
      "totalQuantity": 5,
      "type": "Item",
      "usedQuantity": 5,
      "weekly_purchase_cap": 0,
    },
    Object {
      "buyPrice": 250,
      "buyPriceEach": 10,
      "components": Array [
        Object {
          "buyPrice": 500,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 500,
          "craftResultPrice": 500,
          "daily_purchase_cap": 0,
          "decisionPrice": 500,
          "disciplines": Array [],
          "id": 4,
          "min_rating": null,
          "multipleRecipeCount": 1,
          "output": 1,
          "prerequisites": Array [],
          "quantity": 2,
          "totalQuantity": 50,
          "type": "Item",
          "usedQuantity": 50,
          "weekly_purchase_cap": 0,
        },
      ],
      "craft": true,
      "craftDecisionPrice": 500,
      "craftPrice": 500,
      "craftResultPrice": 500,
      "daily_purchase_cap": 0,
      "decisionPrice": 500,
      "disciplines": Array [],
      "id": 3,
      "min_rating": null,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [],
      "quantity": 5,
      "totalQuantity": 25,
      "type": "Recipe",
      "usedQuantity": 25,
      "weekly_purchase_cap": 0,
    },
    Object {
      "buyPrice": 25000,
      "buyPriceEach": 1000,
      "components": Array [
        Object {
          "buyPrice": 550,
          "buyPriceEach": 11,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 550,
          "craftResultPrice": 550,
          "daily_purchase_cap": 0,
          "decisionPrice": 550,
          "disciplines": Array [],
          "id": 6,
          "min_rating": null,
          "multipleRecipeCount": 1,
          "output": 1,
          "prerequisites": Array [],
          "quantity": 2,
          "totalQuantity": 50,
          "type": "Item",
          "usedQuantity": 50,
          "weekly_purchase_cap": 0,
        },
      ],
      "craft": true,
      "craftDecisionPrice": 550,
      "craftPrice": 550,
      "craftResultPrice": 550,
      "daily_purchase_cap": 0,
      "decisionPrice": 550,
      "disciplines": Array [],
      "id": 5,
      "min_rating": null,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [],
      "quantity": 5,
      "totalQuantity": 25,
      "type": "Recipe",
      "usedQuantity": 25,
      "weekly_purchase_cap": 0,
    },
  ],
  "craft": true,
  "craftDecisionPrice": 1260,
  "craftPrice": 1260,
  "craftResultPrice": 1260,
  "daily_purchase_cap": 0,
  "decisionPrice": 1260,
  "disciplines": Array [
    "Chef",
  ],
  "id": 1,
  "min_rating": 123,
  "multipleRecipeCount": 1,
  "output": 1,
  "prerequisites": Array [],
  "quantity": 1,
  "totalQuantity": 5,
  "type": "Recipe",
  "usedQuantity": 5,
  "weekly_purchase_cap": 0,
}
`;

exports[`updateTree updates a tree correctly if the available items changed 1`] = `
Object {
  "buyPrice": 50,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 0,
      "buyPriceEach": 42,
      "components": undefined,
      "craft": false,
      "craftDecisionPrice": 0,
      "craftResultPrice": 0,
      "daily_purchase_cap": 0,
      "decisionPrice": 0,
      "disciplines": Array [],
      "id": 2,
      "min_rating": null,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [],
      "quantity": 1,
      "totalQuantity": 5,
      "type": "Item",
      "usedQuantity": 0,
      "weekly_purchase_cap": 0,
    },
    Object {
      "buyPrice": 250,
      "buyPriceEach": 10,
      "components": Array [
        Object {
          "buyPrice": 500,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 500,
          "craftResultPrice": 500,
          "daily_purchase_cap": 0,
          "decisionPrice": 500,
          "disciplines": Array [],
          "id": 4,
          "min_rating": null,
          "multipleRecipeCount": 1,
          "output": 1,
          "prerequisites": Array [],
          "quantity": 2,
          "totalQuantity": 50,
          "type": "Item",
          "usedQuantity": 50,
          "weekly_purchase_cap": 0,
        },
      ],
      "craft": true,
      "craftDecisionPrice": 500,
      "craftPrice": 500,
      "craftResultPrice": 500,
      "daily_purchase_cap": 0,
      "decisionPrice": 500,
      "disciplines": Array [],
      "id": 3,
      "min_rating": null,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [],
      "quantity": 5,
      "totalQuantity": 25,
      "type": "Recipe",
      "usedQuantity": 25,
      "weekly_purchase_cap": 0,
    },
    Object {
      "buyPrice": 25000,
      "buyPriceEach": 1000,
      "components": Array [
        Object {
          "buyPrice": 550,
          "buyPriceEach": 11,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 550,
          "craftResultPrice": 550,
          "daily_purchase_cap": 0,
          "decisionPrice": 550,
          "disciplines": Array [],
          "id": 6,
          "min_rating": null,
          "multipleRecipeCount": 1,
          "output": 1,
          "prerequisites": Array [],
          "quantity": 2,
          "totalQuantity": 50,
          "type": "Item",
          "usedQuantity": 50,
          "weekly_purchase_cap": 0,
        },
      ],
      "craft": true,
      "craftDecisionPrice": 550,
      "craftPrice": 550,
      "craftResultPrice": 550,
      "daily_purchase_cap": 0,
      "decisionPrice": 550,
      "disciplines": Array [],
      "id": 5,
      "min_rating": null,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [],
      "quantity": 5,
      "totalQuantity": 25,
      "type": "Recipe",
      "usedQuantity": 25,
      "weekly_purchase_cap": 0,
    },
  ],
  "craft": true,
  "craftDecisionPrice": 1050,
  "craftPrice": 1050,
  "craftResultPrice": 1050,
  "daily_purchase_cap": 0,
  "decisionPrice": 1050,
  "disciplines": Array [
    "Chef",
  ],
  "id": 1,
  "min_rating": 123,
  "multipleRecipeCount": 1,
  "output": 1,
  "prerequisites": Array [],
  "quantity": 1,
  "totalQuantity": 5,
  "type": "Recipe",
  "usedQuantity": 5,
  "weekly_purchase_cap": 0,
}
`;
